boilermaker/app.py

```diff
@@ -74,15 +82,27 @@ async def apply_async(self, fn, *args, delay: int = 0, **kwargs):
         task = self.task_registry[fn.__name__]
         task_copy = copy.deepcopy(task)
         task_copy.payload = payload
-        return await self.publish_task(task_copy, delay=delay)
+        return await self.publish_task(task_copy, delay=delay, retry_policy=retry_policy)
 
     @tracer.start_as_current_span("publish-task")
-    async def publish_task(self, task: Task, delay: int = 0):
+    async def publish_task(self, task: Task, delay: int = 0, retry_policy: RetryPolicy = RetryPolicy.default()):
         """Turn the task into JSON and publish to Service Bus"""
-        await self.service_bus_client.send_message(
-            task.model_dump_json(),
-            delay=delay,
-        )
+        encountered_errors = []
+
+        for i in range(0, retry_policy.max_tries):
+            try:
+                await self.service_bus_client.send_message(
+                    task.model_dump_json(),
+                    delay=delay,
+                )
+                break
+            except Exception as e:
+                encountered_errors.append(e)
+
+                if i == retry_policy.max_tries - 1:
+                    raise BoilermakerAppException("Error encountered while publishing task to service bus", encountered_errors)
+
+            time.sleep(retry_policy.get_delay_interval(i+1))
```

Comment:
This `sleep` is syncronous, so it will block the async event loop. To actually `sleep` we would need `asyncio.sleep`. 

One trick with sleeping and republishing here, though, is that it would potentially greatly delay the actual delivery of the task, which means the _task's_ retry-policy could change its meaning. For instance, the delay_interval max could be violated.

Thus, for _this_ reattempt to publish scenario, I would probably suggest _not_ using a `RetryPolicy`, instead just naively trying 3 or 5 times to publish, but I would also look for explicit messages around connecting to Service Bus as exceptions (instead of `except Exception`). The reason is this:

- we may attempt to publish
- the task may be successfully published, however
- we get some read timeout or something (I'm not sure if this is possible with servicebus; I just wanted to offer it as a thought experiment), so
- the task _has_ published, but we are _also_ going to try again

That too could be a problem.

Reply:
Great points, I'm not sure what I was thinking there! Everything should be updated now, thank you for the great feedback!